### PR TITLE
Refactor reconciliation

### PR DESCRIPTION
**Pull request by galal-hussein**
Regactoring Reconciliation
Add role based add/delete for services

#72 #66

### CODE REVIEW

File: cluster/reconcile.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	if err := reconcileControl(currentCluster, kubeCluster, key, kubeClient); err != nil {
+		return err
+	}
+	logrus.Infof("[reconcile] Reconciled cluster state successfully")
+	return nil
+}
+
+func reconcileWorker(currentCluster, kubeCluster *Cluster, key ssh.Signer, kubeClient *kubernetes.Clientset) error {
+	// worker deleted first to avoid issues when worker+controller on same host
+	logrus.Infof("[reconcile] Check worker hosts to be deleted")
```
Comment (alena1108) [inline cluster/reconcile.go]: I don't think we should log the `check` statements in INFO, may be the debug would be more appropriate

File: cluster/reconcile.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+	}
+
+	kubeClient, err := k8s.NewClient(kubeCluster.LocalKubeConfigPath)
+	if err != nil {
+		return fmt.Errorf("Failed to initialize new kubernetes client: %v", err)
+	}
+
+	key, _ := checkEncryptedKey(kubeCluster.SSHKeyPath)
+
+	if err := reconcileWorker(currentCluster, kubeCluster, key, kubeClient); err != nil {
```
Comment (alena1108) [inline cluster/reconcile.go]: Any particular reason workers are being reconciled first? The official upgrade doc advises to upgrade etcd/master first: https://kubernetes.io/docs/getting-started-guides/ubuntu/upgrades/

File: cluster/reconcile.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+		if err := hosts.DeleteNode(&toDeleteHost, kubeClient, services.ControlRole); err != nil {
+			return fmt.Errorf("Failed to delete worker node %s from cluster", toDeleteHost.Address)
+		}
+		// attempting to clean services/files on the host
+		if err := reconcileHost(toDeleteHost, key, true); err != nil {
+			logrus.Warnf("[reconcile] Couldn't clean up worker node [%s]: %v", toDeleteHost.Address, err)
+			continue
+		}
+	}
+	if err := currentCluster.InvertIndexHosts(); err != nil {
```
Comment (alena1108) [inline cluster/reconcile.go]: I feel like this check belongs to the caller method. Also it is a bit concerning we are calling this check method after delete is already performed. I'd do it as a very first step.

File: docker/docker.go [hunk 4]
```diff
@@ -217,10 +218,10 @@ func IsContainerUpgradable(dClient *client.Client, imageCfg *container.Config, c
 	if err != nil {
 		return false, err
 	}
-	if containerInspect.Config.Image == imageCfg.Image {
-		logrus.Debugf("[%s] Container %s is not eligible for updgrade on host [%s]", plane, containerName, hostname)
-		return false, nil
+	if containerInspect.Config.Image != imageCfg.Image || !reflect.DeepEqual(containerInspect.Config.Cmd, imageCfg.Cmd) {
+		logrus.Debugf("[%s] Container %s is eligible for updgrade on host [%s]", plane, containerName, hostname)
```
Comment (alena1108) [inline docker/docker.go]: Nit: should %s be surrounded with [] for container? to kinda follow up the format you are using

File: hosts/hosts.go [hunk 5]
```diff
@@ -28,17 +28,54 @@ const (
...
 		ToCleanCNIBin,
 		ToCleanCalicoRun,
 	}
+	return h.CleanUp(toCleanPaths)
+}
+
+func (h *Host) CleanUpWorkerHost(controlRole string) error {
+	for _, role := range h.Role {
+		if role == controlRole {
+			logrus.Infof("[hosts] Host [%s] is already a controlplane host, skipping cleanup.", h.Address)
```
Comment (alena1108) [inline hosts/hosts.go]: I'd rephrase this message to something like: `Worker host is a control place host..`

File: services/controlplane.go [hunk 6]
```diff
@@ -9,6 +9,15 @@ import (
 func RunControlPlane(controlHosts []hosts.Host, etcdHosts []hosts.Host, controlServices v1.RKEConfigServices) error {
 	logrus.Infof("[%s] Building up Controller Plane..", ControlRole)
 	for _, host := range controlHosts {
+
+		for _, role := range host.Role {
```
Comment (alena1108) [inline services/controlplane.go]: Looks like we iterate over host roles a lot in the code to figure out if the host is worker vs control. Would it be helpful to have 2 booleans in the host struct - isControl,isWorker - and refer to them instead?

File: cluster/reconcile.go (hunk 2 above)
Comment (galal-hussein) [reply cluster/reconcile.go]: This is for the cases where a host has both controller and worker role, and this same host is the only controller host, when the user run ./rke up it should first remove worker components including kubelet and kubeproxy, if we remove the controller components first it won't be able to remove the worker

Comment (galal-hussein) [reply cluster/reconcile.go]: Also upgrade is not a part of reconcile, its a part of adding components

**Comment by alena1108** [comment]
LGTM
